### ISSUE TITLE
Polymer 2: Multiple CSS mixin definitions mess up each other.

### ISSUE THREAD

**Issue by freshp86** [Polymer/polymer#4486]
When two different parent elements define a mixin for the same child element, the two definitions stomp on each other. This is better explained with the code below

index.html
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="import" href="mixin_bug_elements.html">
  </head>
  <body>
    <container-element></container-element>
    <other-container-element></other-container-element>
  </body>
</html>
```

mixin_bug_elements.html (See comments indicating the bugs).
```html
<link rel="import" href="bower_components/shadycss/apply-shim.html">
<link rel="import" href="bower_components/polymer/polymer-element.html">

<dom-module id="child-element">
  <template>
    <style>
      p {
        @apply --my-mixin;
      }
    </style>
    <p>I'm a DOM element. This is my local DOM!</p>
  </template>
  <script>
    class ChildElement extends Polymer.Element {
      static get is() { return "child-element"; }
    }
    customElements.define(ChildElement.is, ChildElement);
  </script>
</dom-module>

<dom-module id="container-element">
  <template>
    <style>
      child-element {
        --my-mixin: {
          /* Uncomment following line and everything works.*/
          /* font-size: 30px; */
          background-color: yellow;
        }
      }
    </style>
    <child-element></child-element>
  </template>
  <script>
    class ContainerElement extends Polymer.Element {
      static get is() { return "container-element"; }
    }
    customElements.define(ContainerElement.is, ContainerElement);
  </script>
</dom-module>

<dom-module id="other-container-element">
  <template>
    <style>
      child-element {
        --my-mixin: {
          /* font-size is ignored unless the previous  --my-mixin definition also
             contains a font-size rule. */
          font-size: 40px;
          background-color: red;
        }
      }
    </style>
    <child-element></child-element>
  </template>
  <script>
    class OtherContainerElement extends Polymer.Element {
      static get is() { return "other-container-element"; }
    }
    customElements.define(OtherContainerElement.is, OtherContainerElement);
  </script>
</dom-module>
```

Actual result: font-size: 40px on the 2nd instance is ignored.
![mixin_messed_up](https://cloud.githubusercontent.com/assets/7547039/24576610/e07cdf78-1673-11e7-9195-ed3947cddb5b.png)

Un-commenting the font-size rule in the 1st definition of `--my-mixin` causes both instances to be displayed with the correct font-size.
![mixin_working](https://cloud.githubusercontent.com/assets/7547039/24576611/e3e5ddd6-1673-11e7-9b8e-5173fc19bc21.png)

In other words, there is an assumption that all mixin definitions contain the exact same CSS rules. This seems like a bug to me.

**Comment by TomK** [comment]
Related
https://github.com/webcomponents/shadycss/issues/69
https://github.com/Polymer/polymer/issues/4410

**Comment by freshp86** [comment]
I found the problem to be at line https://github.com/webcomponents/shadycss/blob/master/entrypoints/apply-shim.js#L120, where there is no `else` clause. Even though the CSS text has been calculated correctly, it is never applied to the template, because the element has not been registered yet.

https://github.com/freshp86/shadycss/commit/861a52bcb0e7fe728483daca29c760b940bfdc98 fixes the issue for me (but not sure if this is the right fix).

**Comment by azakus** [comment]
Looks like this is a bug in the updating instances with the Apply Shim.

If you force ShadyDOM to be enabled, this works as expected: http://jsbin.com/diyopa/2/edit?html,output

**Comment by freshp86** [comment]
Thanks for confirming. Feel free to experiment. I think my proposed fix at https://github.com/freshp86/shadycss/commit/861a52bcb0e7fe728483daca29c760b940bfdc98, showcases the problem but I don't have enough knowledge about the code to know whether this is the right fix. I am pretty sure though that currently the code does nothing because `shadowRoot` does not exist yet, at the time the code is checking for it at https://github.com/webcomponents/shadycss/blob/master/entrypoints/apply-shim.js#L120.

**Comment by azakus** [comment]
@freshp86 I think I found the root issue. The invalidation system for mixins assumes that there are only two states, valid and invalid. If a template is invalid, we only recalculate the mixin once, and wait a microtask to reset the state. However, `--my-mixin` gets invalidated twice in the same microtask. I think this means I need to keep two counters: `lastVersion` and `currentVersion`, setting `currentVersion` every time a mixin is invalidated, and `lastVersion` after the microtask.

**Comment by azakus** [comment]
Given that this is a Polymer bug, I'll close this issue and reopen the previous one in Polymer. Sorry for the churn!
